Add RzIL chapter and various small changes. (#131)
rizinorg/book: src/disassembling/rzil.md

# RzIL

> RzIL is the new intermediate language in Rizin, primarily intended for representing the semantics of machine code. It is designed as a clone of BAP's [Core Theory](http://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/), with minor deviations where necessary.

Author: maybe i can mention that it resembles lisp?

> More details related the implementation can be found [here](https://github.com/rizinorg/rizin/blob/dev/doc/rzil.md).

Reviewer: ```suggestion
More details related to the implementation can be found [here](https://github.com/rizinorg/rizin/blob/dev/doc/rzil.md).
```

or `relating`

## IL statements

> IL statements for each [opcode](https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_il/rz_il_opcodes.h) can be represented as a string, composed of s-expressions via command `aoi` (JSON output is available via `aoj`).

Reviewer: I am always confused about the term `opcode`. Shouldn't it be `instruction` here?
`opcode` is for the me the bits identifying the instruction.

Author: you are right. the il opcode is like a microcode for an instruction. i will change this.

> In the below shows an example using `aoip` (the *prettified* output) of 2 PPC instructions (`stwu` and `mflr`).

Reviewer: ```suggestion
Here an example using `aoip` (the *prettified* output) on two PPC instructions (`stwu` and `mflr`).
```

```bash
[0x10000488]> pd 2
|           0x10000488      stwu  r1, -0x10(r1)
|           0x1000048c      mflr  r0
[0x10000488]> aoi?
Usage: aoi[p]   # Print the RzIL of next N instructions
| aoi [<n_instructions>]  # Print the RzIL of next N instructions
| aoip [<n_instructions>] # Pretty print the RzIL of next N instructions
[0x10000488]> aoip 2
0x10000488
(seq
  (storew 0
    (+
      (var r1)
      (let v
        (bv 16 0xfff0)
        (ite
          (msb
            (var v))
          (cast 32
            (msb
              (var v))
            (var v))
          (cast 32
            false
            (var v)))))
    (cast 32
      false
      (var r1)))
  (set r1
    (+
      (var r1)
      (let v
        (bv 16 0xfff0)
        (ite
          (msb
            (var v))
          (cast 32
            (msb
              (var v))
            (var v))
          (cast 32
            false
            (var v)))))))
0x1000048c
(set r0
  (cast 32
    false
    (var lr)))
```

> The same output of `aoi` can be obtained via `rz-asm` as in the following example.

Reviewer: ```suggestion
The same output of `aoi` can be obtained via `rz-asm` like this:
```

Just to keep it short.

```bash
$ rz-asm -de -a ppc 7c0802a6
mflr r0
$ rz-asm -Ie -a ppc 7c0802a6
(set r0 (cast 32 false (var lr)))
```

> 

Reviewer 2: Please add that it's possible to print RzIL for the whole function using `plf` command.

Author: oh, i missed this command.

## Emulation

> Rizin enables instruction emulation by leveraging RzIL. This process can record interactions between instructions and VM components like registers and memory. The emulation is controlled via the `aez` commands.

Reviewer: `... record interactions between instructions...``` what does it mean?

```bash
[0x00000000]> aez?
Usage: aez<isv?>   # RzIL Emulation
| aezi                     # Initialize the RzIL Virtual Machine at the current offset
| aezs [<n_times>]         # Step N instructions within the RzIL Virtual Machine
| aezse[j] [<n_times>]     # Step N instructions within the RzIL VM and output VM changes (read &
                             write)
| aezsu <address>          # Step until PC equals given address
| aezsue <address>         # Step until PC equals given address and output VM changes (read & write)
| aezv[jqt] [<var_name> [<number>]] # Print or modify the current status of the RzIL Virtual Machine
```

> Supported architectures can be inspected via `La` command; If the architecture has an `I` as in the example below, then it supports RzIL.

Reviewer: ```suggestion
Supported architectures can be inspected via the `La` command.
If the architecture has an `I`, as in the example below, it supports RzIL.
```

```bash
_dAeI 32 64      ppc         BSD     Capstone PowerPC disassembler
```

> Example of emulation of some instructions in a powerpc binary.

Reviewer: ```suggestion
Example of emulating instructions in a PowerPC binary.
```

> In the example, `r9` contains the base address which is used to calculate the pointer to the string (stored in `r3`) used by 'reloc.printf'.

Reviewer: ```suggestion
In this example, `r9` contains the base address which is used to calculate the pointer to the string (stored in `r3`) used by 'reloc.printf'.
```

# RzIL

> RzIL is the new intermediate language in Rizin, primarily intended for representing the semantics of machine code. It is designed as a clone of BAP's [Core Theory](http://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/), with minor deviations where necessary; it is worth nothing that in practice, RzIL is very similar to the SMT representation with bitvectors and bitvector-indexed arrays as well as effects.

Reviewer 2: "worth noting"

0x10000450 (set r0 (cast 32 false (var lr)))
0x10000454 (seq (storew 0 (+ (var r1) (let v (bv 16 0x14) (ite (msb (var v)) (cast 32 (msb (var v)) (var v)) (cast 32 false (var v))))) (cast 32 false (var r0))) empty)
0x10000458 (seq (set r0 (let ea (+ (var r1) (let v (bv 16 0x14) (ite (msb (var v)) (cast 32 (msb (var v)) (var v)) (cast 32 false (var v))))) (let loadw (loadw 0 32 (var ea)) (cast 32 false (var loadw))))) empty)
0x1000045c (seq (set a (var r1)) (set b (let v (bv 16 0x10) (ite (msb (var v)) (cast 32 (msb (var v)) (var v)) (cast 32 false (var v))))) empty (set r1 (+ (var a) (var b))) empty empty empty)
0x10000460 (set lr (cast 32 false (var r0)))
0x10000464 (seq (set CIA (bv 32 0x10000464)) empty empty (set NIA (& (bv 32 0xfffffffc) (var lr))) (jmp (var NIA)))
[0x100002bc]> 
```

The same output of `aoi` can be obtained via `rz-asm` like this:

```bash
$ rz-asm -de -a ppc 7c0802a6
mflr r0
$ rz-asm -Ie -a ppc 7c0802a6
(set r0 (cast 32 false (var lr)))
```

## Emulation

> Rizin enables instruction emulation by leveraging RzIL. The emulation can be used to record changes within the VM, like read and writes of registers and memory locations. The emulation is controlled via the `aez` commands.

Reviewer: it requires `e io.buffers=true` currently. Otherwise memory read/writes don't work. Should we mention this here?
It took me like two times +20min to figure this out.

Author: yes, we should mention this.
